File: backend/src/main/java/com/back/back9/global/redis/service/RedisService.java
package com.back.back9.global.redis.service;

import com.back.back9.domain.websocket.vo.CandleInterval;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class RedisService {

    private final RedisTemplate<String, String> redisTemplate;
    private static final ObjectMapper mapper = new ObjectMapper();
    private static final int MAX_EXISTING_TO_CHECK = 1000;

    public int saveCandleArray(CandleInterval interval, String symbol, JsonNode candleArray) {
        String redisKey = interval.redisKey(symbol);
        Map<String, JsonNode> existingMap = new HashMap<>();
        int inserted = 0;

        List<String> existing = redisTemplate.opsForList().range(redisKey, 0, MAX_EXISTING_TO_CHECK - 1);
        if (existing != null) {
            for (String json : existing) {
                try {
                    JsonNode parsed = mapper.readTree(json);
                    String time = parsed.path("candle_date_time_kst").asText();
                    existingMap.put(time, parsed);
                } catch (Exception ignored) {
                }
            }
        }

        for (JsonNode candle : candleArray) {
            String kstTime = candle.path("candle_date_time_kst").asText();
            ObjectNode modified = candle.deepCopy();
            modified.remove("candle_date_time_utc");

            long ts = candle.has("timestamp") ? candle.path("timestamp").asLong(0) : 0;
            if (ts == 0 && !kstTime.isEmpty()) {
                try {
                    LocalDateTime ldt = LocalDateTime.parse(kstTime, DateTimeFormatter.ISO_LOCAL_DATE_TIME);
                    ts = ldt.toInstant(ZoneOffset.ofHours(9)).toEpochMilli();
                } catch (Exception e) {
                    log.warn("❗ timestamp 보정 실패: {}", kstTime);
                }
            }
            modified.put("timestamp", ts);

            // 병합 여부
            if (existingMap.containsKey(kstTime)) {
                JsonNode merged = mergeCandles(existingMap.get(kstTime), modified);
                redisTemplate.opsForList().remove(redisKey, 1, existingMap.get(kstTime).toString());
                redisTemplate.opsForList().leftPush(redisKey, merged.toString());
            } else {
                redisTemplate.opsForList().leftPush(redisKey, modified.toString());
            }

            inserted++;
        }

        redisTemplate.opsForList().trim(redisKey, 0, interval.getMaxSize() - 1);
        return inserted;
    }

    public void saveCandle(CandleInterval interval, String symbol, JsonNode candle) {
        String redisKey = interval.redisKey(symbol);
        String kstTime = candle.path("candle_date_time_kst").asText();

        List<String> existing = redisTemplate.opsForList().range(redisKey, 0, MAX_EXISTING_TO_CHECK - 1);
        ObjectNode modified = candle.deepCopy();
        long ts = candle.has("timestamp") ? candle.path("timestamp").asLong(0) : 0;

        if (ts == 0 && !kstTime.isEmpty()) {
            try {
                LocalDateTime ldt = LocalDateTime.parse(kstTime, DateTimeFormatter.ISO_LOCAL_DATE_TIME);
                ts = ldt.toInstant(ZoneOffset.ofHours(9)).toEpochMilli();
            } catch (Exception e) {
                log.warn("❗ WebSocket timestamp 변환 실패: {}", kstTime);
            }
        }
        modified.put("timestamp", ts);

        if (existing != null) {
            for (String json : existing) {
                try {
                    JsonNode parsed = mapper.readTree(json);
                    String time = parsed.path("candle_date_time_kst").asText();
                    if (kstTime.equals(time)) {
                        JsonNode merged = mergeCandles(parsed, modified);
                        redisTemplate.opsForList().remove(redisKey, 1, json);
                        redisTemplate.opsForList().leftPush(redisKey, merged.toString());
                        redisTemplate.opsForList().trim(redisKey, 0, interval.getMaxSize() - 1);
                        log.debug("🔁 캔들 병합 완료 (symbol={}, KST={})", symbol, kstTime);
                        return;
                    }
                } catch (Exception ignored) {}
            }
        }

        redisTemplate.opsForList().leftPush(redisKey, modified.toString());
        redisTemplate.opsForList().trim(redisKey, 0, interval.getMaxSize() - 1);
        log.debug("✅ 캔들 저장 완료 (symbol={}, KST={}, timestamp={})", symbol, kstTime, ts);
    }

    public void saveLatestCandle(String symbol, JsonNode candleNode) {
        String key = symbol + ":Latest";
        redisTemplate.opsForValue().set(key, candleNode.toString());
    }

    private JsonNode mergeCandles(JsonNode oldCandle, JsonNode newCandle) {
        ObjectNode merged = oldCandle.deepCopy();

        // 시가(opening_price): 더 이른 시간의 값
        long oldTs = oldCandle.path("timestamp").asLong(0);
        long newTs = newCandle.path("timestamp").asLong(0);
        merged.put("opening_price", oldTs <= newTs ? oldCandle.path("opening_price").asDouble() : newCandle.path("opening_price").asDouble());

        // 종가(trade_price): 더 늦은 시간의 값
        merged.put("trade_price", oldTs >= newTs ? oldCandle.path("trade_price").asDouble() : newCandle.path("trade_price").asDouble());

        // 고가(high_price): 최대
        merged.put("high_price", Math.max(oldCandle.path("high_price").asDouble(), newCandle.path("high_price").asDouble()));

        // 저가(low_price): 최소
        merged.put("low_price", Math.min(oldCandle.path("low_price").asDouble(), newCandle.path("low_price").asDouble()));

        // 누적 거래량
        merged.put("candle_acc_trade_volume", oldCandle.path("candle_acc_trade_volume").asDouble(0) + newCandle.path("candle_acc_trade_volume").asDouble(0));

        // 누적 거래금액
        merged.put("candle_acc_trade_price", oldCandle.path("candle_acc_trade_price").asDouble(0) + newCandle.path("candle_acc_trade_price").asDouble(0));

        return merged;
    }

    public List<JsonNode> getLatestCandle(CandleInterval interval, String symbol) {
        String key = interval.redisKey(symbol);
        List<String> rawList = redisTemplate.opsForList().range(key, 0, 169);
        if (rawList == null) return List.of();

        return rawList.stream().map(this::parseJson).sorted(Comparator.comparing(node -> LocalDateTime.parse(node.path("candle_date_time_kst").asText()), Comparator.reverseOrder())).collect(Collectors.toList());
    }

    public List<JsonNode> getPreviousCandlesByRange(CandleInterval interval, String symbol, int page, LocalDateTime time) {
        final int PAGE_SIZE = 50;
        final int INITIAL_RENDER_SIZE = 120;
        final int OFFSET = INITIAL_RENDER_SIZE + (page * PAGE_SIZE);
        final int REDIS_FETCH_LIMIT = 1000;

        String key = interval.redisKey(symbol);
        List<String> rawList = redisTemplate.opsForList().range(key, 0, REDIS_FETCH_LIMIT - 1);
        if (rawList == null || rawList.isEmpty()) return List.of();

        List<JsonNode> parsedList = rawList.stream().map(this::parseJson).filter(node -> {
            String dateStr = node.path("candle_date_time_kst").asText();
            try {
                return LocalDateTime.parse(dateStr).isBefore(time);
            } catch (Exception e) {
                return false;
            }
        }).sorted(Comparator.comparing(node -> LocalDateTime.parse(node.path("candle_date_time_kst").asText()), Comparator.reverseOrder())).toList();

        int fromIndex = Math.min(OFFSET, parsedList.size());
        int toIndex = Math.min(fromIndex + PAGE_SIZE, parsedList.size());

        return parsedList.subList(fromIndex, toIndex);
    }

    public JsonNode getLatest1sCandle(String symbol) {
        String key = symbol + ":Latest";
        String value = redisTemplate.opsForValue().get(key);
        return (value == null) ? null : parseJson(value);
    }

    private JsonNode parseJson(String str) {
        try {
            return mapper.readTree(str);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("JSON 파싱 실패", e);
        }
    }

    public int countCandles(CandleInterval interval, String symbol) {
        String key = interval.redisKey(symbol);
        Long size = redisTemplate.opsForList().size(key);
        return size != null ? size.intValue() : 0;
    }

    public void clearAll() {
        Optional.ofNullable(redisTemplate.getConnectionFactory()).map(RedisConnectionFactory::getConnection).ifPresent(connection -> {
            try (connection) {
                connection.serverCommands().flushDb();
            }
        });
    }

    public void sortAndRewrite() {
        log.info("✅ Redis key 전체 정렬 시작");

        Set<String> keys = redisTemplate.keys("*");
        if (keys == null) return;

        for (String key : keys) {
            List<String> list = redisTemplate.opsForList().range(key, 0, -1);
            if (list == null || list.isEmpty()) continue;

            try {
                List<Map<String, Object>> parsedList = new ArrayList<>();
                for (String item : list) {
                    parsedList.add(mapper.readValue(item, Map.class));
                }

                Map<String, Map<String, Object>> deduped = new TreeMap<>();
                for (Map<String, Object> entry : parsedList) {
                    String time = (String) entry.get("candle_date_time_kst");
                    deduped.put(time, entry);
                }

                List<String> sorted = deduped.values().stream().sorted(Comparator.comparing(e -> (String) e.get("candle_date_time_kst"))).map(e -> {
                    try {
                        return mapper.writeValueAsString(e);
                    } catch (Exception ex) {
                        return null;
                    }
                }).filter(Objects::nonNull).toList();

                redisTemplate.delete(key);
                for (String json : sorted) {
                    redisTemplate.opsForList().rightPush(key, json);
                }

                log.info("✅ Redis key '{}' 정렬 완료 ({}개)", key, sorted.size());

            } catch (Exception e) {
                log.error("❌ Redis 정렬 중 오류 발생 (key={}): {}", key, e.getMessage());
            }
        }
    }
}

File: backend/src/main/kotlin/com/back/back9/global/redis/service/RedisService.kt
package com.back.back9.global.redis.service

import com.back.back9.domain.websocket.vo.CandleInterval
import com.back.back9.global.redis.dto.Candle
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service

@Service
class RedisService(
    private val redisTemplate: RedisTemplate<String, String>,
    private val objectMapper: ObjectMapper
) {

    private val log = LoggerFactory.getLogger(RedisService::class.java)

    /** 캔들 데이터를 Redis의 Sorted Set에 저장합니다. score는 candle의 timestamp입니다. */
    fun saveCandle(key: String, candle: Candle) {
        try {
            val candleJson = objectMapper.writeValueAsString(candle)
            redisTemplate.opsForZSet().add(key, candleJson, candle.timestamp.toDouble())
        } catch (e: JsonProcessingException) {
            log.error("Candle 객체를 JSON으로 변환하는 데 실패했습니다. Key: {}, Candle: {}", key, candle, e)
        }
    }

    /** JsonNode 형태의 캔들 데이터를 저장합니다. */
    fun saveCandle(interval: CandleInterval, market: String, candleNode: JsonNode) {
        try {
            val candle = objectMapper.treeToValue(candleNode, Candle::class.java)
            saveCandle(interval.redisKey(market), candle)
        } catch (e: JsonProcessingException) {
            log.error("JsonNode를 Candle 객체로 변환하는 데 실패했습니다. Market: {}, Node: {}", market, candleNode, e)
        }
    }

    /** 최신 캔들을 저장합니다. */
    fun saveLatestCandle(market: String, candle: Candle) {
        val key = CandleInterval.SEC.redisKey(market)
        saveCandle(key, candle)
    }

    /** 최신 캔들을 JsonNode로 저장합니다. */
    fun saveLatestCandle(market: String, candleNode: JsonNode) {
        try {
            val candle = objectMapper.treeToValue(candleNode, Candle::class.java)
            saveLatestCandle(market, candle)
        } catch (e: JsonProcessingException) {
            log.error("JsonNode를 Candle 객체로 변환하는 데 실패했습니다. Market: {}, Node: {}", market, candleNode, e)
        }
    }

    /** 지정된 키에서 최신 'count'개의 캔들을 조회합니다. */
    fun getLatestCandles(key: String, count: Int): List<Candle> {
        val candleJsonSet = redisTemplate.opsForZSet()
            .reverseRange(key, 0, count.toLong() - 1)
        return parseCandleSet(key, candleJsonSet)
    }

    /** 커서(timestamp) 이전의 캔들을 조회합니다. */
    fun getCandlesBefore(key: String, cursorTimestamp: Long, count: Int): List<Candle> {
        val candleJsonSet = redisTemplate.opsForZSet()
            .reverseRangeByScore(key, Double.NEGATIVE_INFINITY, (cursorTimestamp - 1).toDouble(), 0, count.toLong())
        return parseCandleSet(key, candleJsonSet)
    }

    /** 가장 최신 캔들 1개를 조회합니다. */
    fun getLatestCandle(key: String): Candle? {
        val candleJson = redisTemplate.opsForZSet()
            .reverseRange(key, 0, 0)
            ?.firstOrNull() ?: return null
        return try {
            objectMapper.readValue(candleJson, Candle::class.java)
        } catch (e: JsonProcessingException) {
            log.error("JSON을 Candle 객체로 변환하는 데 실패했습니다. Key: {}, JSON: {}", key, candleJson, e)
            null
        }
    }

    /** JSON Set을 Candle 리스트로 변환합니다. */
    private fun parseCandleSet(key: String, candleJsonSet: Set<String>?): List<Candle> {
        if (candleJsonSet.isNullOrEmpty()) return emptyList()
        return candleJsonSet.mapNotNull { json ->
            try {
                objectMapper.readValue(json, Candle::class.java)
            } catch (e: JsonProcessingException) {
                log.error("JSON을 Candle 객체로 변환하는 데 실패했습니다. Key: {}, JSON: {}", key, json, e)
                null
            }
        }.sortedByDescending { it.timestamp }
    }

    /** 캔들 개수를 반환합니다. */
    fun countCandles(key: String): Long =
        redisTemplate.opsForZSet().zCard(key) ?: 0L

    /** Redis의 모든 키를 삭제합니다. */
    fun clearAll() {
        val keys = redisTemplate.keys("*")
        if (keys.isNotEmpty()) {
            redisTemplate.delete(keys)
            log.info("Redis의 모든 데이터를 삭제했습니다. (총 {}개 키)", keys.size)
        }
    }

    /** 특정 timestamp 이전의 데이터를 조회 후 삭제합니다. */
    fun getAndRemoveCandlesBefore(key: String, endTimestamp: Long): List<Candle> {
        val minScore = Double.NEGATIVE_INFINITY
        val maxScore = endTimestamp.toDouble()
        val candleJsonSet = redisTemplate.opsForZSet().rangeByScore(key, minScore, maxScore)
            ?: return emptyList()

        redisTemplate.opsForZSet().removeRangeByScore(key, minScore, maxScore)

        log.info("Key '{}'에서 {} 이전 데이터 {}개를 백업하고 삭제했습니다.", key, endTimestamp, candleJsonSet.size)

        return parseCandleSet(key, candleJsonSet)
    }
}
